### PR TITLE
150-monai-flexibility

### LINKED ISSUE #150: Tutorial showcasing modularity of MONAI and how it can be integrated into your existing training workflow.

### LINKED ISSUE #151: Tutorials showcasing different use-cases/networks (segmentation, classification) etc. using MONAI end-end

### PR DESCRIPTION
**Pull request by madil90**
Fixes #151 .

### Description
Example that shows how MONAI components (like transforms) can be used in an existing pytorch workflow. Picked the [60-min blitz pytorch example](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) and added transforms to it. 

Also adds 2D densenet to example and show performance increase in few extra lines of code.

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: examples/cifar10_to_monai.ipynb [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "env: CUDA_VISIBLE_DEVICES=1\n"
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: No need to set CUDA visible device in this tutorial, if user only has 1 GPU, it may cause an error to set =1.

File: examples/cifar10_to_monai.ipynb [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "\n",
+    "# functions to show an image\n",
+    "\n",
+    "def imshow(img):\n",
+    "    img = img / 2 + 0.5     # unnormalize\n",
+    "    npimg = img.numpy()\n",
+    "    plt.imshow(np.transpose(npimg, (1, 2, 0)))\n",
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: Could you please change to use **plt.figure()** and **plt.subplot()** to draw a bigger picture with fixed size?
And just use **plt.title("XXX")** to set label name, your labels are not aligned with images.

File: examples/cifar10_to_monai.ipynb [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+      "Epoch 2/2, Iteration 10000/12500: Loss = 1.3130395187735557\n",
+      "Epoch 2/2, Iteration 12000/12500: Loss = 1.2754147632271051\n",
+      "Finished Training in 117.91511368751526 seconds.\n"
+     ]
+    }
+   ],
+   "source": [
+    "from time import time\n",
+    "\n",
+    "no_epochs = 2\n",
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: Why call it "**no_epochs**"? I think maybe "epochs" or "num_epochs" is more clear.

File: examples/cifar10_to_monai.ipynb [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+      "Predicted:    cat  ship  ship  ship\n"
+     ]
+    }
+   ],
+   "source": [
+    "dataiter = iter(testloader)\n",
+    "images, labels = dataiter.next()\n",
+    "\n",
+    "# print images\n",
+    "imshow(torchvision.utils.make_grid(images))\n",
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: Same as the preview logic. plt.subplot() should be better.

File: examples/cifar10_to_monai.ipynb [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "cuda:0\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(device)"
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: duplicated operation, already print out in previous logics.

File: examples/cifar10_to_monai.ipynb [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accuracy of the network on the 10000 test images: 52 %\n"
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: The accuracy looks very low, should we train more epochs?

File: examples/cifar10_to_monai.ipynb [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accuracy of the network on the 10000 test images: 56 %\n"
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: Seems the result is similar after you added several random transforms from MONAI, should we show these transforms obviously help the training?

File: examples/cifar10_to_monai.ipynb [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+    "transform = transforms.Compose(\n",
+    "                [PILToNumpy(),\n",
+    "                 monai.transforms.AsChannelFirst(),\n",
+    "                 monai.transforms.RandFlip(prob=0.5, spatial_axis=[1, 0]),\n",
+    "                 monai.transforms.RandRotate(prob=0.5, degrees=90),\n",
+    "                 monai.transforms.IntensityNormalizer(np.array((0, 0, 0)), \n",
+    "                                                      np.array((255, 255, 255))),\n",
+    "                 monai.transforms.GaussianNoise(),\n",
+    "                 monai.transforms.ToTensor(),\n",
+    "                 transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])"
```
Comment (Nic-Ma) [inline examples/cifar10_to_monai.ipynb]: Why you add **Normalize** again as you already added "IntensityNormalizer"? I think they are the same logic.

File: examples/cifar10_to_monai.ipynb (hunk 1 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: makes sense. i'll fix this.

File: examples/cifar10_to_monai.ipynb (hunk 2 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: This is part of the original cifar-10 tutorial. Kept it like this so users can easily recognize.

File: examples/cifar10_to_monai.ipynb (hunk 3 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: Original tutorial called it no_epochs. Also no_ is common alternative of num_

File: examples/cifar10_to_monai.ipynb (hunk 4 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: Will use same resolution as above.

File: examples/cifar10_to_monai.ipynb (hunk 5 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: Will remove this. Thanks.

File: examples/cifar10_to_monai.ipynb (hunk 6 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: Yes. changed to more epochs.

File: examples/cifar10_to_monai.ipynb (hunk 7 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: The only main addition is a random flip transform. CIFAR-10 is a very simple dataset. To show big accuracy increase, will need a more complicated dataset.

File: examples/cifar10_to_monai.ipynb (hunk 8 above)
Comment (madil90) [reply examples/cifar10_to_monai.ipynb]: IntensityNormalizer works on images and Normalize on tensors. One is to get image to 0-1 range. Other is to convert output to -0.5-0.5 range. Explained this.

File: examples/cifar10_to_monai.ipynb (hunk 2 above)
Comment (Nic-Ma) [reply examples/cifar10_to_monai.ipynb]: OK, let's keep it.

File: examples/cifar10_to_monai.ipynb (hunk 3 above)
Comment (Nic-Ma) [reply examples/cifar10_to_monai.ipynb]: I didn't find `no_epoch` in the original tutorial, just:
```py
for epoch in range(2):
    ... ...
```

File: examples/cifar10_to_monai.ipynb (hunk 4 above)
Comment (Nic-Ma) [reply examples/cifar10_to_monai.ipynb]: OK, let's keep it.

File: examples/cifar10_to_monai.ipynb (hunk 8 above)
Comment (Nic-Ma) [reply examples/cifar10_to_monai.ipynb]: OK, actually you can set (0.5, 0.5, 0.5) to `IntensityNormalizer` and no need to call `Normalize`.
But it's better to show how we compose with torchvision transforms together, so let's keep it.
`IntensityNormalizer` had been changed name to `NormalizeIntensity` in the latest code, please merge master and update your tutorial.

**Comment by wyli** [comment]
as discussed in the meeting we should consider MedNIST Dataset https://courses.nvidia.com/courses/course-v1:DLI+L-HX-07+V1/about

**Comment by Nic-Ma** [comment]
Close the PR as we plan to use MedNIST to develop another tutorial.
Thanks.
